### PR TITLE
Configure TLS for private v2 registry mirrors.

### LINKED ISSUE #14823: Docker 1.8 does not configure TLS for private registry mirrors
Fix in progress.  Please assign to me.

cc @stevvooe @dmp42 @tiborvass

### PR DESCRIPTION
**Pull request by RichardScothern**
Signed-off-by: Richard Scothern richard.scothern@gmail.com

Closes #14823 

cc @tiborvass @dmcgowan

### CODE REVIEW

File: registry/service.go [hunk 1]
```diff
@@ -161,19 +162,35 @@ func (s *Service) TlsConfig(hostname string) (*tls.Config, error) {
 	return &tlsConfig, nil
 }
 
+func (s *Service) tlsConfigForMirror(mirror string) (*tls.Config, error) {
+	mirrorUrl, err := url.Parse(mirror)
+	if err != nil {
+		return &tls.Config{}, err
+	}
+	tlsConfig, err := s.TlsConfig(mirrorUrl.Host)
```
Comment (tiborvass) [inline registry/service.go]: you can simply `return s.TlsConfig(...)`

File: registry/service.go [hunk 2]
```diff
@@ -161,19 +162,35 @@ func (s *Service) TlsConfig(hostname string) (*tls.Config, error) {
 	return &tlsConfig, nil
 }
 
+func (s *Service) tlsConfigForMirror(mirror string) (*tls.Config, error) {
+	mirrorUrl, err := url.Parse(mirror)
+	if err != nil {
+		return &tls.Config{}, err
```
Comment (calavera) [inline registry/service.go]: this can be `return nil, err`

File: registry/service.go [hunk 3]
```diff
@@ -161,19 +162,35 @@ func (s *Service) TlsConfig(hostname string) (*tls.Config, error) {
...
+	}
+	tlsConfig, err := s.TlsConfig(mirrorUrl.Host)
+	if err != nil {
+		return &tls.Config{}, err
```
Comment (calavera) [inline registry/service.go]: this can be `return nil, err`

File: registry/service.go (hunk 1 above)
Comment (calavera) [inline registry/service.go]: agreed with this.

**Comment by stevvooe** [comment]
LGTM after fixing the error returns.

**Comment by tiborvass** [comment]
I wonder what happens with v1 mirrors

**Comment by dmcgowan** [comment]
Can you clarify what the original behavior was that was causing problems?

**Comment by tiborvass** [comment]
@dmcgowan the original problem was that the tls certs were not set up for v2 mirrors.

**Comment by calavera** [comment]
without calling the `s.TlsConfig` function for each mirror you could try to use certificates for other host incorrectly.

**Comment by dmcgowan** [comment]
@tiborvass to clarify my understanding, the issue is that it was not using certificates from `/etc/docker/cert.d` for mirrors, which could include needed client certificates. It should not have been using the wrong certificates since it would have defaulted to the base configuration.

**Comment by dmcgowan** [comment]
> I wonder what happens with v1 mirrors

@tiborvass To your comment above the section, I don't think v1 mirrors should even be included in that list. V1 mirrors do not mirror the index and those endpoints should only be indexes.

**Comment by tiborvass** [comment]
@RichardScothern feel free to remove v1 mirrors in that LookupEndpoints list.

**Comment by dmcgowan** [comment]
LGTM

**Comment by tiborvass** [comment]
Let's make it clearer in the commit message and PR description that this is fixing *private* v2 mirrors and public v2 mirrors were working just fine.

**Comment by tiborvass** [comment]
LGTM

**Pull request merged**